### ISSUE TITLE
Downsampling operation in tensorflow does not have its gradient implemented :o

### ISSUE THREAD

**Issue by tobias-liaudat** [CosmoStat/wf-psf#16]
The operation is the one defined by open-cv INTER_AREA method that is the same as tensorflow's AREA method from [tf.image.resize](https://www.tensorflow.org/api_docs/python/tf/image/resize).

- [open-cv doc](https://docs.opencv.org/2.4/modules/imgproc/doc/geometric_transformations.html#void%20resize(InputArray%20src,%20OutputArray%20dst,%20Size%20dsize,%20double%20fx,%20double%20fy,%20int%20interpolation)).
- [tensorflow doc](https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/resize-area)
- Clarifying medium post on the method [here](https://medium.com/@wenrudong/what-is-opencvs-inter-area-actually-doing-282a626a09b3)

Need to find a workaround for this and to see how we can propagate the gradient through this operation.

**Comment by tobias-liaudat** [comment]
## Proposed solution

Restrain ourselves to integer values for downsampling, and also to values that are divisors of the image dimensions.
We ask `output_Q` to be an integer and to `input_dim/output_Q` to be an integer too.

The first condition might make us choose `output_Q=3` in order to be Nyquist sampled _at all frequencies_ in the VIS passband. If we choose  `output_Q=2` we are not going to be Nyquist sampled for a small range of frequencies _inside_ the Euclid's passband.

The second condition is not a problem as we can choose the size of the `input_dim` when we crop the image on the diffraction operator.

If these two conditions are met we can replace the `tf.image.resize` function by an [AveragePooling2D](https://www.tensorflow.org/api_docs/python/tf/keras/layers/AveragePooling2D) layer. The operation is the same and the gradients are defined.


### Testing method

The results are matching when the conditions are met and the hyper parameters are correctly set as we can see below.

<img width="807" alt="Screenshot 2021-03-31 at 15 15 29" src="https://user-images.githubusercontent.com/16106854/113151681-9f643a80-9235-11eb-9b21-5cbd27e982d5.png">

The notebook `testing-downsample` contains the code to reproduce the test.

### If conditions are not met

The methods differ as we can see below.

<img width="817" alt="Screenshot 2021-03-31 at 15 29 15" src="https://user-images.githubusercontent.com/16106854/113152002-f407b580-9235-11eb-9cbe-f58964d3589e.png">

**Comment by tobias-liaudat** [comment]
This is done and the models are training so I'll close it.
